src/simple_http.rs

```diff
@@ -184,54 +204,51 @@ impl Builder {
-
-            // We need to get the hostname and the port.
-            // (1) Split scheme
-            let after_scheme = {
-                let mut split = url_cloned.splitn(2, "://");
-                let s = split.next().unwrap();
-                split.next().unwrap_or(s)
-            };
-            // (2) split off path
-            let before_path = after_scheme.splitn(2, "/").next().unwrap();
-            // (3) split off auth part
-            let after_auth = {
-                let mut split = before_path.splitn(2, "@");
-                let s = split.next().unwrap();
-                split.next().unwrap_or(s)
-            };
-            // so now we should have <hostname>:<port> or just <hostname>
-            let mut split = after_auth.split(":");
-            let hostname = split.next().unwrap();
-            let port: u16 = match split.next() {
-                Some(port_str) => match port_str.parse() {
-                    Ok(port) => port,
-                    Err(_) => return Err(Error::InvalidUrl(url)),
-                },
-                None => DEFAULT_PORT,
-            };
-            // make sure we don't have a second colon in this part
-            if split.next().is_some() {
-                return Err(Error::InvalidUrl(url));
+        // We need to get the hostname and the port.
+        // (1) Split scheme
+        let after_scheme = {
+            let mut split = url.splitn(2, "://");
+            let s = split.next().unwrap();
+            split.next().unwrap_or(s)
+        };
+        // (2) split off path
+        let (before_path, path) = {
+            if let Some(slash) = after_scheme.find("/") {
+                (&after_scheme[0..slash], &after_scheme[slash..])
+            } else {
+                (after_scheme, "/")
             }
-
-            self.tp.addr = match format!("{}:{}", hostname, port).to_socket_addrs()?.next() {
-                Some(a) => a,
-                None => return Err(Error::InvalidUrl(url)),
-            };
+        };
+        // (3) split off auth part
+        let after_auth = {
+            let mut split = before_path.splitn(2, "@");
+            let s = split.next().unwrap();
+            split.next().unwrap_or(s)
+        };
+        // so now we should have <hostname>:<port> or just <hostname>
+        let mut split = after_auth.split(":");
+        let hostname = split.next().unwrap();
+        let port: u16 = match split.next() {
+            Some(port_str) => match port_str.parse() {
+                Ok(port) => port,
+                Err(_) => return Err(Error::InvalidUrl(url.to_owned())),
+            },
+            None => DEFAULT_PORT,
+        };
+        // make sure we don't have a second colon in this part
+        if split.next().is_some() {
+            return Err(Error::InvalidUrl(url.to_owned()));
         }
 
-        self.tp.url = url;
+        self.tp.addr = match format!("{}:{}", hostname, port).to_socket_addrs()?.next() {
```

Comment:
There should be no need for `format!` here, `ToSocketAddrs` is implemented for `(&str, u16)`

Reply:
fixing